compaction/compaction.cc

```diff
@@ -1703,6 +1704,98 @@ future<uint64_t> scrub_validate_mode_validate_reader(flat_mutation_reader_v2 rea
     co_return errors;
 }
 
+future<uint64_t> scrub_validate_mode_validate_index(sstables::shared_sstable sst, reader_permit permit, const io_priority_class& pc, const compaction_data& cdata) {
+    uint64_t errors{0};
+    auto idx_reader = sstables::index_reader(sst, permit, pc, {}, sstables::use_caching::no);
+    const auto& schema = *sst->get_schema();
+    partition_key::tri_compare cmp(schema);
+    std::optional<partition_key> last_pkey;
+    std::optional<uint64_t> last_pos;
+
+    std::exception_ptr ex;
+    try {
+        while (!idx_reader.eof()) {
+            if (cdata.is_stop_requested()) [[unlikely]] {
+                // Compaction manager will catch this exception and re-schedule the compaction.
+                throw compaction_stopped_exception(schema.ks_name(), schema.cf_name(), cdata.stop_requested);
+            }
+
+            co_await idx_reader.read_partition_data();
+            auto pos = idx_reader.get_data_file_position();
+            auto pkey = idx_reader.get_partition_key();
+
+            if (last_pkey && cmp(*last_pkey, pkey) <= 0) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Out-of-order partition {} ({}) in index, previous partition is {} ({})",
+                        schema.ks_name(),
+                        schema.cf_name(),
+                        scrub_compaction::partition_key_to_string(pkey, schema),
+                        pkey,
+                        scrub_compaction::partition_key_to_string(*last_pkey, schema),
+                        *last_pkey);
+            }
+
+            if (last_pos && pos <= *last_pos) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Out-of-order position {} for partition {} ({}) in index, previous position is {} for partition {} ({})",
+                        schema.ks_name(),
+                        schema.cf_name(),
+                        pos,
+                        scrub_compaction::partition_key_to_string(pkey, schema),
+                        pkey,
+                        *last_pos,
+                        scrub_compaction::partition_key_to_string(*last_pkey, schema),
+                        *last_pkey);
+            }
+
+            auto pr = dht::partition_range::make_singular(dht::decorate_key(schema, pkey));
+            auto reader = sst->make_reader(sst->get_schema(), permit, pr, schema.full_slice(), pc, {}, streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
```

Comment:
Won't this be super slow with tiny partitions? Also, this uses the index which is being validated.

If we're validating the entire sstable, we can co-scan the data file and index and compare them along.

Reply:
Yes it will be very slow.

I guess it will be better to do this on the low-level indeed.
I thought I can get away with reading the index + fast-forwarding. This proved to be a dead-end quickly. Recreating the reader indeed is very icky. I forgot about the fact that if the index is corrupt the lookup might simply fail, even though the data is there.